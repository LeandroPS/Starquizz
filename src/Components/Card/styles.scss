@import '../../styles/common';
@import '../../styles/colors';

.card{
    width: 100%;
    border-radius: 5px;
    overflow: hidden;

    transition: all 0.3s cubic-bezier(.25,.8,.25,1);
    position: relative;
    top: 0;
    left: 0;
    transform: translate(0%, 0%);

    @extend %box-shadow;

    &:hover{
        @extend %box-shadow-hovered;
    }

    &.guessed{

        button{
            display: none;
        }
    }

    &.grows{
        position: fixed;
        top: 50%;
        left: 50%;
        transform: translate(-50%, -50%);

        width: 500px;
        margin: 0;

        transition: all 1s;
    }

    .thumb{
        padding-bottom: 100%;

        background-color: $gray;
        background-size: cover;
        background-position: top center;
        background-repeat: no-repeat;

        button{
            position: absolute;
            top: 10px;
            right: 10px;
        }
    }

    .answer{
        position: absolute;
        height: 60px;
        background-color: #fff;
        width: 100%;
        padding: 10px;
        box-sizing: border-box;
        bottom: 0;
        transition: bottom 1s;

        input{
            outline: none;
            -webkit-appearance: none;
            border: none;
            background-color: transparent;
            font: inherit;
            height: 40px;

            &:disabled{
                font-weight: bold;
                color: $black;
            }
        }

        button{
            position: absolute;
            right: 10px;
        }
    }

    button{
        @extend .button; 
    }
}