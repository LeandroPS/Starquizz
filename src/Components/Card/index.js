import React, {Component} from 'react';
import PropTypes from 'prop-types';

import CharacterWindow from '../CharacterWindow';
import axios from 'axios';

import './styles.scss';

class Card extends Component{

    constructor(){
        super();

        this.state = {
            showAnwerField: false,
            showCharacterWindow: false,
            usedHelp: false,
            value: "",
            image: null
        }
    }

    componentDidMount(){
        this.getImage(this.props.character.name);
    }

    componentDidUpdate(prevProps){
        if(prevProps.character.name !== this.props.character.name) this.getImage(this.props.character.name);
    }

    getImage = (term) =>{
        let APIKey = "AIzaSyCJndh_Z04YbVnhZ3s6TA0QcVzQ5RX_PQY",
            CustomSearchID = "010694824975899644089:m6l68ukm8ma",
            base_url = "https://www.googleapis.com/customsearch/v1";

        axios.get(base_url,{
            params:{
                key: APIKey,
                cx: CustomSearchID,
                searchType: 'image',
                q: term
            }
         })
            .then((response)=>{
                this.setState({image: response.data.items[0].link});
            });
    }

    onChange = (value) => {
        this.setState({ value });
    }

    onGetHelp = () => {
        this.setState({usedHelp: true, showCharacterWindow: true});
    }

    onCloseHelpModal = () => {
        this.setState({showCharacterWindow: false});
    }

    onAnswer = (answer) => {
        const {character: {name}} = this.props;

        if(answer === name) this.onRightAnswer();
    }

    onRightAnswer = () => {
        const {usedHelp} = this.state;

        this.props.includeGuessedCharacter(this.props.character.name);
        this.props.addPoints(usedHelp?5:10);
    }

    render(){

        const {showAnwerField, showCharacterWindow, value} = this.state;
        const {character, guessed, getResource} = this.props;

        return(
            <React.Fragment>
                <div className={`card ${guessed&&'guessed'}`}>
                    <div className="thumb" style={{backgroundImage: `url(${this.state.image})`}}>
                        {!guessed&&(
                            <button onClick={()=>this.onGetHelp()}>
                                <i className="fas fa-question"></i>
                            </button>
                        )}
        
                    </div>
                    <div className={`answer ${showAnwerField&&"show"}`}>
                        <input 
                            type="text" 
                            placeholder="Resposta"
                            value={value}
                            onChange={(e)=>this.onChange(e.target.value)}
                            disabled={guessed}/>
                        {!guessed&&(
                            <button
                                onClick={()=>this.onAnswer(value)}>
                                Ok
                            </button>
                        )}
                    </div>
                </div>
                {showCharacterWindow&&
                    (<CharacterWindow 
                        character={{...character, ...{image: this.state.image}}}
                        onClose={this.onCloseHelpModal}
                        getResource={getResource}/>)
                }
            </React.Fragment>
        )
    }
}

Card.propTypes = {
    character: PropTypes.object.isRequired, 
    guessed: PropTypes.bool.isRequired, 
    getResource: PropTypes.func.isRequired,
    includeGuessedCharacter: PropTypes.func.isRequired,
    addPoints: PropTypes.func.isRequired
};

export default Card;