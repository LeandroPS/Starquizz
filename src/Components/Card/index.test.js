import React from 'react';
import ReactDOM from 'react-dom';
import { create } from "react-test-renderer";
import { shallow, mount } from 'enzyme';
import Card from './';

describe('Card', () => {
    const props = {
        character: {
            species: [],
            name: 'Luke Skywalker',
            height: 123,
            homeworld: 'earth',
            films: [],
            vehicles: []
        },  
        guessed: false,
        getResource: jest.fn(),
        includeGuessedCharacter: jest.fn(),
        addPoints: jest.fn()
    };

    it('renders without crashing', () => {
        const div = document.createElement('div');
        ReactDOM.render(<Card {...props}/>, div);
        ReactDOM.unmountComponentAtNode(div);
    });

    it('renders correctly', () => {
        const tree = create(<Card {...props}/>)
          .toJSON();
        expect(tree).toMatchSnapshot();
    });

    it('open details window when clicked help button', () => {
        const component = mount(<Card {...props}/>);
        const button = component.find(".thumb button")
        button.simulate("click");
        
        expect(component.find("div.character-window").length).toBe(1);
    });

    it('call includeGuessedCharacter on right answer', () => {
        const component = shallow(<Card {...props}/>);
        const input  = component.find(".answer input");
        input.simulate('change', { target: { value: 'Luke Skywalker' } });
        component.find(".answer button").simulate("click");
        expect(props.includeGuessedCharacter).toHaveBeenCalled();
    });

});

