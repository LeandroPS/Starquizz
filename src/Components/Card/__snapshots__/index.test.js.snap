// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card renders correctly 1`] = `
<div
  className="card false"
>
  <div
    className="thumb"
    style={
      Object {
        "backgroundImage": "url(null)",
      }
    }
  >
    <button
      onClick={[Function]}
    >
      <i
        className="fas fa-question"
      />
    </button>
  </div>
  <div
    className="answer false"
  >
    <input
      disabled={false}
      onChange={[Function]}
      placeholder="Resposta"
      type="text"
      value=""
    />
    <button
      onClick={[Function]}
    >
      Ok
    </button>
  </div>
</div>
`;
